charmcraft/providers.py

```diff
@@ -105,6 +109,66 @@ def get_command_environment() -> Dict[str, str]:
     return env
 
 
+@contextlib.contextmanager
+def launch_environment(
+    *,
+    charm_name: str,
+    project_path: pathlib.Path,
+    base: Base,
+    bases_index: int,
+    build_on_index: int,
+    lxd_project: str = "charmcraft",
+    lxd_remote: str = "local",
+):
+    """Launch environment for specified base.
+
+    :param charm_name: Name of project.
+    :param project_path: Path to project.
+    :param base: Base to create.
+    :param bases_index: Index of `bases:` entry.
+    :param build_on_index: Index of `build-on` within bases entry.
+    """
+    alias = BASE_CHANNEL_TO_BUILDD_IMAGE_ALIAS[base.channel]
+    target_arch = get_host_architecture()
+
+    instance_name = get_instance_name(
+        bases_index=bases_index,
+        build_on_index=build_on_index,
+        project_name=charm_name,
+        target_arch=target_arch,
+    )
+
+    environment = get_command_environment()
+    image_remote = configure_buildd_image_remote()
+    base_configuration = CharmcraftBuilddBaseConfiguration(
+        alias=alias, environment=environment, hostname=instance_name
+    )
+    instance = lxd.launch(
+        name=instance_name,
+        base_configuration=base_configuration,
+        image_name=base.channel,
+        image_remote=image_remote,
+        auto_clean=True,
+        auto_create_project=True,
+        map_user_uid=True,
+        use_snapshots=True,
+        project=lxd_project,
+        remote=lxd_remote,
+    )
+
+    # Mount project.
+    instance.mount(
+        host_source=project_path, target=get_managed_environment_project_path()
+    )
+
+    yield instance
+
+    # Unmount everything.
+    instance.unmount_all()
```

Comment:
if instance raises, this won't be run, neither will stop, maybe add a test for that.

Reply:
Ohh, good point.  I'll put a finally clause here and add a test.

Reply:
Done!